htmresearch/frameworks/clustering/sdr_clustering.py

```diff
@@ -142,31 +148,82 @@ def addCluster(self, cluster):
+  def _removeCluster(self, cluster):
+    clusterId = cluster.getId()
+    if clusterId in self._clusters:
+      self._clusters.pop(clusterId)
+      _LOGGER.debug('REMOVED: cluster with ID %s was removed' % clusterId)
+
 
-    clusterDistPairs = self.computeClusterDistances(cluster)
-    notMerged = True
+  def _mergeNewCluster(self):
+
+    clusterDistPairs = computeClusterDistances(self._newCluster, 
+                                               self.getClusters())
+    clusterMerged = False
     if len(clusterDistPairs) > 0:
-      for clusterDistPair in clusterDistPairs:
-        closestClusterToNewDist, closestClusterToNew = clusterDistPair
+      closestClusterDist, closestCluster = clusterDistPairs[0]
+      if closestClusterDist < self._mergeThreshold:
+        _LOGGER.debug("MERGE: Cluster %s merged with cluster %s. "
+                      "Inter-cluster distance: %s"
+                      % (self._newCluster.getId(),
+                         closestCluster.getId(),
+                         closestClusterDist))
+        updateTime = self._numPointsClustered
+        closestCluster.merge(self._newCluster, updateTime)
+        clusterMerged = True
+
+    return clusterMerged
 
-        if closestClusterToNewDist < self._mergeThreshold:
-          _LOGGER.debug("Cluster %s merged with cluster %s. Inter-cluster "
-                        "distance: %s" % (cluster.getId(),
-                                          closestClusterToNew.getId(),
-                                          closestClusterToNewDist))
-          closestClusterToNew.merge(cluster)
-          notMerged = False
 
-    if notMerged:
-      self.addCluster(cluster)
+  def _mergeCluster(self, cluster, clusters):
+    """
+    Merge cluster if it is close enough to a cluster in the list and return 
+    True. Otherwise, don't merge the cluster and return False.
+    :param cluster: (Cluster) cluster to merge w/ one of the existing clusters. 
+    :param clusters: (list of Clusters) list of clusters to compare the first 
+    cluster to.
+    :return: (bool) Wether or not the cluster was merged.
+    """
+
+    clusterDistPairs = computeClusterDistances(cluster, clusters)
+    clusterMerged = False
+    if len(clusterDistPairs) > 0:
+      for clusterDistPair in clusterDistPairs:
+        closestClusterDist, closestCluster = clusterDistPair
+
+        if closestClusterDist < self._mergeThreshold:
+          _LOGGER.debug("MERGE: Cluster %s merged with cluster %s. "
+                        "Inter-cluster distance: %s" % (cluster.getId(),
+                                                        closestCluster.getId(),
+                                                        closestClusterDist))
+          closestCluster.merge(cluster, self._numPointsClustered)
+          self._removeCluster(cluster)
+          clusterMerged = True
+
+    return clusterMerged
+
+
+  def _pruneClusters(self):
+    clusters = self.getClusters()
+    for i in range(len(clusters)):
```

Comment:
As per our discussion, since we don't need and don't use pruning at the moment, I marked it as not implemented. If we do use pruning in the future, we should iterate through all permutations of remaining clusters to make sure we are not missing a cluster that should actually be merged.